controller_server:
  ros__parameters:
    controller_frequency: 5.0
    costmap_update_timeout: 0.30
    min_x_velocity_threshold: 0.001
    min_y_velocity_threshold: 0.001 
    min_theta_velocity_threshold: 0.001
    failure_tolerance: 0.3
    odom_topic: "/ona2/odometry"
    progress_checker_plugins: ["progress_checker"]
    goal_checker_plugins: ["general_goal_checker"] # "precise_goal_checker"
    controller_plugins: ["FollowPath"]
    use_realtime_priority: false

    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.5
      movement_time_allowance: 10.0

    general_goal_checker:
      stateful: True
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.35
      yaw_goal_tolerance: 0.50

    FollowPath:
      plugin: "mappi::MPPIcROS"

      GeneralSettings:
        transform_tolerance: 0.5
        num_iterations: 1
        batch_size: 1000
        time_steps: 100
        num_retry: 4
        control_offset: 1
        goal_tolerance: 0.5
        plan_shift: 1.0
        use_splines: false
        model_dt: 0.05
        temperature: 0.3
        gamma: 0.05

      Constraints:
        max_vx: 0.75
        min_vx: -0.75
        max_vy: 0.5
        min_vy: -0.5
        max_wz: 0.5
        min_wz: -0.5
        max_ax: 1.0
        min_ax: -1.0
        max_ay: 1.0
        min_ay: -1.0
        max_az: 2.5

      MotionModel: "BicycleKin"

      Ackermann:
        min_radius: 3.0

      BicycleKin:
        length: 1.5
        length_rear: 0.75
        max_steer: 0.3

      NoiseGeneration:
        std_vx: 0.3
        std_vy: 0.3
        std_wz: 0.1

      Critics:
        Goal:
          active: true
          power: 1
          weight: 1.0
          threshold: 1.5
        GoalAngle:
          active: true
          power: 1
          weight: 1.0
          threshold: 1.5
        PathDist:
          active: false
          power: 2
          weight: 5.0
          threshold: 1.0
          stride: 2
        Twirling:
          active: false
          power: 1
          weight: 10.0
        Forward:
          active: false
          power: 1
          weight: 10.0
        PathFollow:
          active: true
          power: 1
          weight: 10.0
          threshold: 1.0
          offset_from_furthest: 3
        PathAngle:
          active: false
          power: 1
          weight: 25.0
          threshold: 1.0
          offset_from_furthest: 3
          angle_threshold: 90.0
        Obstacles:
          active: true
          power: 1
          weight: 50.0
          threshold: 1.0
          repulsive_weight: 0.0
          collision_cost: 100000.0
          collision_margin_dist: 0.2

      Visualization:
        active: true
        batch_stride: 100
        time_stride: 3
        default_z: 0.2
        scale: 0.05